File: Матюхин/1722.ts
function dfs(
    globalReachArray: boolean[],
    reachArray: boolean[],
    vertex: number,
    edges: number[][],
): void {
    reachArray = reachArray.fill(false);
    const dfsStack = [vertex];
    globalReachArray[vertex] = true;
    reachArray[vertex] = true;

    while (dfsStack.length !== 0) {
        const currentVertex = dfsStack.pop();

        edges.forEach((edge) => {
            if (edge[0] === currentVertex && !reachArray[edge[1]]) {
                globalReachArray[edge[1]] = true;
                reachArray[edge[1]] = true;
                dfsStack.push(edge[1]);
            }

            if (edge[1] === currentVertex && !reachArray[edge[0]]) {
                globalReachArray[edge[0]] = true;
                reachArray[edge[0]] = true;
                dfsStack.push(edge[0]);
            }
        })
    }
}

function minimumHammingDistance(
    source: number[],
    target: number[],
    allowedSwaps: number[][],
): number {
    let distance = source.length;
    const globalReachArray = new Array(source.length).fill(false);
    const reachArray = new Array(source.length);
    const reachedSource = new Array(source.length);
    const reachedTarget = new Array(target.length);

    for (let i = 0; i < source.length; ++i) {
        if (globalReachArray[i]) {
            continue;
        }

        dfs(
            globalReachArray,
            reachArray,
            i,
            allowedSwaps,
        );

        let reachedSourceIndex = 0;
        let reachedTargetIndex = 0;
        let sameElementsNumber = 0;
        reachedSource.length = 0;
        reachedTarget.length = 0;

        for (let j = 0; j < source.length; ++j) {
            if (reachArray[j]) {
                reachedSource[reachedSource.length] = source[j];
                reachedTarget[reachedTarget.length] = target[j];
            }
        }

        reachedSource.sort((x, y) => x - y);
        reachedTarget.sort((x, y) => x - y);

        while (
            reachedSourceIndex < reachedSource.length &&
            reachedTargetIndex < reachedTarget.length
        ) {
            if (
                reachedSource[reachedSourceIndex] <
                reachedTarget[reachedTargetIndex]
            ) {
                ++reachedSourceIndex;
            } else if (
                reachedSource[reachedSourceIndex] >
                reachedTarget[reachedTargetIndex]
            ) {
                ++reachedTargetIndex;
            } else {
                ++sameElementsNumber;
                ++reachedSourceIndex;
                ++reachedTargetIndex;
            }
        }

        distance -= sameElementsNumber;
    }

    return distance;
}

File: Матюхин/1722.go
import "sort"

func dfs(
    globalReachArray []bool,
    reached *[]int,
    dfsStack []int,
    vertex int,
    adjacentVertices []map[int]bool,
) {
    dfsStackLength := 1
    dfsStack[0] = vertex
    globalReachArray[vertex] = true

    *reached = (*reached)[:0]
    *reached = append(*reached, vertex)

    for dfsStackLength > 0 {
        currentVertex := dfsStack[dfsStackLength - 1]
        dfsStackLength--

        for adjacentVertex := range adjacentVertices[currentVertex] {
            if !globalReachArray[adjacentVertex] {
                globalReachArray[adjacentVertex] = true
                *reached = append(*reached, adjacentVertex)
                dfsStack[dfsStackLength] = adjacentVertex
                dfsStackLength++
            }
        }
    }
}

func minimumHammingDistance(
    source []int,
    target []int,
    allowedSwaps [][]int,
) int {
    distance := len(source)
    adjacentVertices := make([]map[int]bool, len(source))
    dfsStack := make([]int, len(source))
    globalReachArray := make([]bool, len(source))
    reached := make([]int, len(source))
    reachedSource := make([]int, len(source))
    reachedTarget := make([]int, len(source))

    for mapKey := range adjacentVertices {
        adjacentVertices[mapKey] = make(map[int]bool)
    }

    for _, allowedSwap := range allowedSwaps {
        adjacentVertices[allowedSwap[0]][allowedSwap[1]] = true
        adjacentVertices[allowedSwap[1]][allowedSwap[0]] = true
    }

    for i := 0; i < len(source); i++ {
        if globalReachArray[i] {
            continue
        }

        dfs(globalReachArray, &reached, dfsStack, i, adjacentVertices)

        reachedSourceIndex := 0
        reachedTargetIndex := 0
        sameElementsNumber := 0
        reachedLength := 0

        for _, j := range reached {
            reachedSource[reachedLength] = source[j]
            reachedTarget[reachedLength] = target[j]
            reachedLength++
        }

        sort.Ints(reachedSource[:reachedLength])
        sort.Ints(reachedTarget[:reachedLength])

        for reachedSourceIndex < reachedLength &&
                reachedTargetIndex < reachedLength {
            if reachedSource[reachedSourceIndex] <
                    reachedTarget[reachedTargetIndex] {
                reachedSourceIndex++
            } else if reachedSource[reachedSourceIndex] >
                    reachedTarget[reachedTargetIndex] {
                reachedTargetIndex++
            } else {
                sameElementsNumber++
                reachedSourceIndex++
                reachedTargetIndex++
            }
        }

        distance -= sameElementsNumber
    }

    return distance
}
